### PR TITLE
shell: Redesign untrusted "add host" dialog

### PR DESCRIPTION
**Pull request by martinpitt**
Remote machines can send malicious cockpit pages which do iframe traversal or use the `host` channel option to run arbitrary code on all other currently connected Cockpit sessions.

Isolating cockpit sessions from different hosts from one another is difficult, and properly implementing this will take a while (if it's possible at all): https://issues.redhat.com/browse/COCKPIT-870

Until this happens, warn users about that in the "unknown host" shell dialog. 

Also hide the "verify fingerprint" recipe behind an expander. In practice, most users do "Trust On First Use" anyway, as validating a FP properly is too hard. This makes the dialog less overwhelming and also puts more emphasis on the "trust" aspect.

Add a check to TestMultiMachine.testTroubleshooting which actually runs the generated validation command and ensures that the shown fingerprint matches.

----

Initial dialog, FP validation collapsed by default:

image

FP validation expanded:

image

FP popover:

image

### CODE REVIEW

**Comment by martinpitt** [comment]
If possible, I'd also like to trim the alert text a bit, to fit on one line (at least in English). Perhaps:

> Malicious remote Web Console pages can (influence|change|damage) other connected hosts in this session.

?

**Comment by garrett** [comment]
This modal is way overloaded. It's not clear what's going on, what's important, what someone needs to do, etc. I'm trying to figure this out in a mockup right now.

As for the new part of this PR, the security issue: It's not error, nor is it an alert (of something that's been done already). Opting in is clicking the button; it doesn't need to have an "unbreak the UI" toggle.

Still, this modal was already too complex before adding anything else, and it's extremely overloaded now. I hope to finish up a suggestion with a mockup soon and I'll share it here when I have it.

**Comment by garrett** [comment]
@martinpitt: Ah, that's what I thought, but wasn't sure. Additionally, corps should really be using CA with SSH certs, but in practice, few do.

https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/deployment_guide/sec-creating_ssh_ca_certificate_signing-keys

https://smallstep.com/blog/use-ssh-certificates/

I guess I'll redesign the modal to be even simpler now!

**Comment by garrett** [comment]
This isn't done, but I thuoght I should share what I have before the upcoming long weekend. I'd like to get feedback.

HOSTNAME is the actual hostname or IP address.

image

I don't have the "View details" popover for the details for the session stuff yet. We might want to have a popover for "certificate authority" and we might want to ditch the extended sharing info.

What do you think so far?

**Comment by garrett** [comment]
Oh, the warning text at the bottom has a typo. I revised it a few times and accidentally left it in that state. :sweat_smile:

**Comment by martinpitt** [comment]
Nice, thanks @garrett ! I like the decomposition/expanders. I'm less sure about the "not using certificate authority"  -- that's one way to assert a known host, but (unfortunately) more of a fringe case, and confusing for casual users. I like the  original "you are connecting to hostname for the first time", as when you see it, it's exactly right, and doesn't induce fear.

**Comment by martinpitt** [comment]
@garrett First iteration towards your design. Initial dialog (not expanded):

image

Expanding:

image

Opening popup:

image

Diff between design and implementation:
 - Does not speak about "certificate authority", see above. I kept the original "for the first time" text, as that's the 95% case.
 - Drop the redundant "Verify the fingerprint by [running command..]", as the heading already says that.

File: pkg/shell/hosts_dialog.jsx [hunk 1]
```diff
@@ -479,18 +484,27 @@ class HostKey extends React.Component {
...
+                <ExpandableSection toggleText={ _("Verify fingerprint") }
+                                   isExpanded={this.state.verifyExpanded}
+                                   onToggle={(_ev, verifyExpanded) => this.setState({ verifyExpanded }) }>
+                    <p>{_("Run this command over a trusted network or physically on the remote machine:")}</p>
+                    <ClipboardCopy isReadOnly hoverTip={_("Copy")} clickTip={_("Copied")} className="hostkey-verify-help pf-v5-u-font-family-monospace">ssh-keyscan -t {key_type} localhost | ssh-keygen -lf -</ClipboardCopy>
+                    <p>{_("The fingerprint should match:")} {fingerprint_help}</p>
+                    <ClipboardCopy isReadOnly hoverTip={_("Copy")} clickTip={_("Copied")} className="hostkey-verify-help pf-v5-u-font-family-monospace">{fp}</ClipboardCopy>
+                </ExpandableSection>
+                <Alert variant='warning' isInline isPlain title={_("Malicious pages on the remote machine can do arbitrary changes to other connected hosts in the same session.")} />
```
Comment (allisonkarlitskaya) [inline pkg/shell/hosts_dialog.jsx:507]: `s/can do/can make/`

**Comment by garrett** [comment]
These changes work for me! 

This seems very verbose and hard to follow:

> Malicious pages on the remote machine can do arbitrary changes to other connected hosts in the same session.

What I shared:

> Pages from the remote machine can _make changes_ to other hosts in the same session. 

(Emphasis for the part I accidentally left out in the mockup. Whoops!)

What about this?

_Malicious pages on a remote system may affect other connected hosts._

File: pkg/shell/hosts_dialog.jsx (hunk 1 above)
Comment (martinpitt) [reply pkg/shell/hosts_dialog.jsx:507]: Superseded by Garrett's proposal below.

**Comment by martinpitt** [comment]
Thanks @garrett and @allisonkarlitskaya ! I updated the warning string, and put current screenshots into the description. Triggering f38 tests to see what I need to adjust.

File: pkg/shell/hosts_dialog.jsx [hunk 2]
```diff
@@ -479,18 +484,27 @@ class HostKey extends React.Component {
                 <p>{cockpit.format(_("To verify a fingerprint, run the following on $0 while physically sitting at the machine or through a trusted network:"), this.props.host)}</p>
                 <ClipboardCopy isReadOnly hoverTip={_("Copy")} clickTip={_("Copied")} className="hostkey-verify-help-cmds pf-v5-u-font-family-monospace">ssh-keyscan -t {key_type} localhost | ssh-keygen -lf -</ClipboardCopy>
                 <p>{_("The resulting fingerprint is fine to share via public methods, including email.")}</p>
-                <p>{_("If the fingerprint matches, click 'Accept key and connect'. Otherwise, do not connect and contact your administrator.")}</p>
+                <p>{_("If the fingerprint matches, click 'Trust and add host'. Otherwise, do not connect and contact your administrator.")}</p>
```
Comment (cockpituous) [inline pkg/shell/hosts_dialog.jsx:487]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19409-20231010-153219-d8bb31b4-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/hosts_dialog.jsx [hunk 3]
```diff
@@ -479,18 +484,27 @@ class HostKey extends React.Component {
...
-                <ClipboardCopy isReadOnly hoverTip={_("Copy")} clickTip={_("Copied")} className="hostkey-fingerprint pf-v5-u-font-family-monospace">{fp}</ClipboardCopy>
-                <p className="hostkey-type">({key_type})</p>
-                <p>{cockpit.format(_("To verify a fingerprint, run the following on $0 while physically sitting at the machine or through a trusted network:"), this.props.host)}</p>
-                <ClipboardCopy isReadOnly hoverTip={_("Copy")} clickTip={_("Copied")} className="hostkey-verify-help-cmds pf-v5-u-font-family-monospace">ssh-keyscan -t {key_type} localhost | ssh-keygen -lf -</ClipboardCopy>
-                <p>{_("The resulting fingerprint is fine to share via public methods, including email.")}</p>
-                <p>{_("If the fingerprint matches, click 'Accept key and connect'. Otherwise, do not connect and contact your administrator.")}</p>
+                <ExpandableSection toggleText={ _("Verify fingerprint") }
+                                   isExpanded={this.state.verifyExpanded}
+                                   onToggle={(_ev, verifyExpanded) => this.setState({ verifyExpanded }) }>
```
Comment (cockpituous) [inline pkg/shell/hosts_dialog.jsx:501]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19409-20231010-153219-d8bb31b4-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/hosts_dialog.jsx (hunk 2 above)
Comment (martinpitt) [reply pkg/shell/hosts_dialog.jsx:487]: I added test for this now.

File: pkg/shell/hosts_dialog.jsx (hunk 3 above)
Comment (martinpitt) [reply pkg/shell/hosts_dialog.jsx:501]: Added a test.

**Comment by martinpitt** [comment]
This is now ready. I did the final tweaks discussed with @garrett (drop the period in the alert), adjusted tests, and cleaned them up a bit while I was at it.

File: pkg/shell/hosts_dialog.jsx [hunk 4]
```diff
@@ -479,18 +484,24 @@ class HostKey extends React.Component {
                 <p>{cockpit.format(_("To verify a fingerprint, run the following on $0 while physically sitting at the machine or through a trusted network:"), this.props.host)}</p>
                 <ClipboardCopy isReadOnly hoverTip={_("Copy")} clickTip={_("Copied")} className="hostkey-verify-help-cmds pf-v5-u-font-family-monospace">ssh-keyscan -t {key_type} localhost | ssh-keygen -lf -</ClipboardCopy>
                 <p>{_("The resulting fingerprint is fine to share via public methods, including email.")}</p>
-                <p>{_("If the fingerprint matches, click 'Accept key and connect'. Otherwise, do not connect and contact your administrator.")}</p>
+                <p>{_("If the fingerprint matches, click 'Trust and add host'. Otherwise, do not connect and contact your administrator.")}</p>
```
Comment (cockpituous) [inline pkg/shell/hosts_dialog.jsx:487]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19409-20231011-054038-2287fc87-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by jelly — APPROVED** [review]
Nice improvements!

**Review by garrett — APPROVED** [review]
Thanks! It's a lot better now! :+1: Ready to merge. :tada:
